File: 윤장원/210629/오픈채팅방.py
import collections

def solution(record):
    answer = []
    hash_table = collections.defaultdict(str)
    record = [r.split() for r in record]
    
    # 최종적으로 바뀐 아이디
    for r in record:
        attr = r[0] 
        u_id = r[1]
        if attr == 'Enter':
            hash_table[u_id] = r[2]
        elif attr == 'Change':
            hash_table[u_id] = r[2]
            
    # record 확인 
    for r in record:
        attr = r[0]
        u_id = r[1]
        if attr == 'Enter':
            answer.append(f'{hash_table[u_id]}님이 들어왔습니다.')
        elif attr == 'Leave':
            answer.append(f'{hash_table[u_id]}님이 나갔습니다.')
            
    return answer

File: 윤장원/210629/오픈채팅방.js
function solution(record) {
    const records = [];
    const hash = new Map();
    const result = [];
    
    // record 데이터 파싱
    for (const r of record) {
        records.push(r.split(' '));
    }
    
    for (const record of records) {
        const attr = record[0];
        const id = record[1];
        const char = record[2];
        
        // 최종적인 아이디만 확인
        if (attr === 'Enter') {
            hash.set(id, char);
        } else if (attr === 'Change') {
            hash.set(id, char);
        }
    }
    
    // 최종적인 로그 확인
    for (const record of records) {
        const attr = record[0];
        const id = record[1];
        const char = record[2];
        
        if (attr === 'Enter') {
            result.push(`${hash.get(id)}님이 들어왔습니다.`);
        } else if (attr === 'Leave') {
            result.push(`${hash.get(id)}님이 나갔습니다.`)
        }
    }
    
    return result;
}